commit 7c80548dba38a170ba33c964d3f6ad8dba49db4f
Author: John Spray <jspray@redhat.com>
Date:   Thu Mar 15 12:28:17 2018 +0000

    mgr: improve reporting on unloadable modules
    
    Previously, a module that failed to load at all (for
    example due to an import error at module scope) would
    be missing from the reports to the monitor, and would
    result in enigmatic messages about a module not being
    present on all mgrs.
    
    Fixes: http://tracker.ceph.com/issues/23358
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/PyModule.cc b/src/mgr/PyModule.cc
index 0aacaf8a9d0..53ca84c703e 100644
--- a/src/mgr/PyModule.cc
+++ b/src/mgr/PyModule.cc
@@ -70,6 +70,24 @@ std::string handle_pyerror()
     return extract<std::string>(formatted);
 }
 
+/**
+ * Get the single-line exception message, without clearing any
+ * exception state.
+ */
+std::string peek_pyerror()
+{
+  PyObject *ptype, *pvalue, *ptraceback;
+  PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+  assert(ptype);
+  assert(pvalue);
+  PyObject *pvalue_str = PyObject_Str(pvalue);
+  std::string exc_msg = PyString_AsString(pvalue_str);
+  Py_DECREF(pvalue_str);
+  PyErr_Restore(ptype, pvalue, ptraceback);
+
+  return exc_msg;
+}
+
 
 namespace {
   PyObject* log_write(PyObject*, PyObject* args) {
@@ -288,10 +306,8 @@ int PyModule::load(PyThreadState *pMainThreadState)
 
     r = load_commands();
     if (r != 0) {
-      std::ostringstream oss;
-      oss << "Missing COMMAND attribute in module '" << module_name << "'";
-      error_string = oss.str();
-      derr << oss.str() << dendl;
+      derr << "Missing COMMAND attr in module '" << module_name << "'" << dendl;
+      error_string = "Missing COMMAND attribute";
       return r;
     }
 
@@ -320,6 +336,7 @@ int PyModule::load(PyThreadState *pMainThreadState)
         if (!PyBool_Check(pCanRun) || !PyString_Check(can_run_str)) {
           derr << "Module " << get_name()
                << " returned wrong type in can_run" << dendl;
+          error_string = "wrong type returned from can_run";
           can_run = false;
         } else {
           can_run = (pCanRun == Py_True);
@@ -333,6 +350,7 @@ int PyModule::load(PyThreadState *pMainThreadState)
       } else {
         derr << "Module " << get_name()
              << " returned wrong type in can_run" << dendl;
+        error_string = "wrong type returned from can_run";
         can_run = false;
       }
 
@@ -410,6 +428,7 @@ int PyModule::load_subclass_of(const char* base_class, PyObject** py_class)
   // load the base class
   PyObject *mgr_module = PyImport_ImportModule("mgr_module");
   if (!mgr_module) {
+    error_string = peek_pyerror();
     derr << "Module not found: 'mgr_module'" << dendl;
     derr << handle_pyerror() << dendl;
     return -EINVAL;
@@ -417,6 +436,7 @@ int PyModule::load_subclass_of(const char* base_class, PyObject** py_class)
   auto mgr_module_type = PyObject_GetAttrString(mgr_module, base_class);
   Py_DECREF(mgr_module);
   if (!mgr_module_type) {
+    error_string = peek_pyerror();
     derr << "Unable to import MgrModule from mgr_module" << dendl;
     derr << handle_pyerror() << dendl;
     return -EINVAL;
@@ -425,6 +445,7 @@ int PyModule::load_subclass_of(const char* base_class, PyObject** py_class)
   // find the sub class
   PyObject *plugin_module = PyImport_ImportModule(module_name.c_str());
   if (!plugin_module) {
+    error_string = peek_pyerror();
     derr << "Module not found: '" << module_name << "'" << dendl;
     derr << handle_pyerror() << dendl;
     return -ENOENT;
diff --git a/src/mgr/PyModule.h b/src/mgr/PyModule.h
index b077beafbbe..df79e8e9a6d 100644
--- a/src/mgr/PyModule.h
+++ b/src/mgr/PyModule.h
@@ -22,6 +22,8 @@
 
 std::string handle_pyerror();
 
+std::string peek_pyerror();
+
 /**
  * A Ceph CLI command description provided from a Python module
  */
